### PR TITLE
Assignment5+6

### PR DESCRIPTION
**Pull request by TechGlue**
Initial PR.
Pair Programming Partner - Nicholas Johnson @NicholasJohnsonDev

### CODE REVIEW

**Comment by MarkCapka** [comment]
Doing code review: mark Capka - 2.23.22 will submit as I go along

**Comment by ZackWard-student** [comment]
Code review in progress

**Comment by MarkCapka** [comment]
Work in progress - Mark Capka 2.23.22 5:15pm

Quick debrief: 
good work, guys! Overall, looking pretty solid. 
comments were left with //NOTE where I could

highlights: 
1. update your analyzers and make sure they are consistent for the Assignment.csproj and Assignment.Tests.csproj
2. hardcoded spokane addresses test
3.ChildItems method for Node
4. selecting using ToArray
5. sorting addresses by multiple parameters
6. unit test cleanup 

But overall, good job! Most of your linq statements were concise and achieved what you were looking for. Tests are largely good as well


Checklist:
 Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔//NOTE: confirmed

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔ //NOTE: confirmed
Using LINQ, skip the first row in the People.csv. ✔//NOTE: confirmed
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔//NOTE: confirmed
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔//NOTE: confirmed - although test implementation suggestion below: 
      //NOTE/Suggestion: you use foreach on this method to iterate through the List<string>, maybe try making an IEnumerable and then selecting the values that you are looking to filter. We were also kind of confused on the testing of this list and ended up using a Mock to compare the results at the end.

Use ISampleData.CsvRows for your data source. ✔ //confirmed 
Don't forget the list should be unique. ✔ //confirmed
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of Spokane-based addresses. ❌
      //NOTE: not included 
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔  NOTE: see suggestion above though, test could be improved I think. 
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔ //confirmed
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ❌ 
   //NOTE: You didn't approach it this way, however, it does seem to work

Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔ //Confirmed

Use ISampleData.CsvRows as the source of the data. ✔  //confirmed 
Sort the list by State, City, Zip. (Sort the addresses first then select). ❌ 
    //NOTE: currently only ordering by State it seems. Can use .ThenBy([nextparam].... and chain those together. 
Be sure that Person.Address is also populated. ✔ //Confirmed
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔ //confirmed 

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma separated list of states. ✔ //confirmed

Use the people parameter from ISampleData.People property for your data source. ✔
 //confirmed
At a minimum, use System.Linq.Enumerable.Aggregate LINQ method to create your result. ✔
  //confirmed 
Don't forget the list should be unique. ✔ //confirmed
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
//confirmed


Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
    //Note: it seems that this is technically implemented and tested as you do point to the head, however, changing name in tests and the node class to include "circle" would help
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.❌
  // Note: missing the ChildItems it seems, 
   

**Extra Credit** //NOTE: not attempted
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals
Ensure you enable:
nullable reference types is enabled ✔ NOT ENABLED in .csproj but specified in project properties
net6 targeted ✔     NOT CURRENTLY SPECIFIED IN .CSPROJ but specified in project properties
C# 10.0 ✔     NOT CURRENTLY SPECIFIED IN .CSPROJ but did work so probably is targetting
and enabled .NET analyzers for both projects ❌ 
             //NOTE: only the INTELLITECT analyzer is currently included. 
For this assignment, favor using Assert.AreEqual() (the generic version) ✔
           //NOTE: generally follows this, several other kinds of asserts to consider whether or not you should use boolean assertion or AreEquals

All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Assignment/Directory.Build.props [hunk 1]
```diff
@@ -1,9 +1,19 @@
 <Project>
-<PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
-    <Nullable>enable</Nullable>
+	<PropertyGroup>
+		<TargetFramework>net6.0</TargetFramework>
+		<LangVersion>10.0</LangVersion>
```
Comment (NicholasJohnsonDev) [inline Assignment/Directory.Build.props:4]: @MarkCapka 
C# 10.0 ✔ NOT CURRENTLY SPECIFIED IN .CSPROJ but did work so probably is targetting
and enabled .NET analyzers for both projects ❌

Is that not what is we are doing here? I think this file works at the solution level but I could be wrong.

File: Assignment/Directory.Build.props [hunk 2]
```diff
@@ -1,9 +1,19 @@
...
+		<ImplicitUsings>enable</ImplicitUsings>
+		<Nullable>enable</Nullable>
+
+		<AnalysisMode>AllEnabledByDefault</AnalysisMode>
+		<EnableNETAnalyzers>true</EnableNETAnalyzers>
+		<!-- 
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <EnableNETAnalyzers>true</EnableNETAnalyzers>
     <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
```
Comment (NicholasJohnsonDev) [inline Assignment/Directory.Build.props:12]: @TechGlue @NicholasJohnsonDev 
Note to remove comments out warnings to errors.

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,43 @@
...
+        {
+            IEnumerable<string> states = GetUniqueSortedListOfStatesGivenCsvRows();
+            return string.Join(", ", states);
+        }
 
         // 4.
-        public IEnumerable<IPerson> People => throw new NotImplementedException();
+        public IEnumerable<IPerson> People => CsvRows
+            .Select(x => x.Split(","))
+            .OrderBy(x => x[6]).ThenBy(x => x[5]).ThenBy(x => x[7])
```
Comment (NicholasJohnsonDev) [inline Assignment/Assignment/SampleData.cs:29]: @MarkCapka 
Sort the list by State, City, Zip. (Sort the addresses first then select). ❌

Am I not doing that on line 29? LINQ is new to me and could be doing it wrong.

**Comment by NicholasJohnsonDev** [comment]
Thank you for your feedback and for catching what we overlooked/misunderstood.

**Comment by ZackWard-student** [comment]
1. Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
> Just missing adequate testing. Try thinking about the tests as a comparison to the expected results.
>I think it feels like the tests are written to match the code.

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of Spokane-based addresses. ❌
> Already noted by programmers
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌
>interesting test but I agree that it does not meet the requirement

3. Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
>Note: using a comma and space in the join might be wrong. They asked for a comma joined. I assumed that was like the csv file. Though you are consistent, so I think you are fine.

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔

> I think you can call .join on a string list instead of calling it string.Join(str, list)

4. Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
>Your code would be more readable if the x's where named what you expect them to be.

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional. (Not used)
Consider using ISampleData.CsvRows in your test to verify your results. ✔

5. Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔

6. Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use System.Linq.Enumerable.Aggregate LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

7. Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ❌
>missing the IEnumerator IEnumerable.GetEnumerator() implementation. try yield returning the implemented one. It might need a cast.
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum. ❌
>This method is missing


Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals
Ensure you enable:
nullable reference types is enabled ✔
net6 targeted ✔
C# 10.0 ✔

>Found these in a Solution Directory  file

and enabled .NET analyzers for both projects ✔
For this assignment, favor using Assert.AreEqual() (the generic version) ✔
All of the above should be unit tested ❌

>missing the appropriate tests for ISampleData.CsvRows. Already >noted
>NOTE: Tests should be written with the expected value on the left
Choose simplicity over complexity ✔

File: Assignment/Assignment/SampleData.cs (hunk 3 above)
Comment (ZackWard-student) [reply Assignment/Assignment/SampleData.cs:29]: I think you are good here.

File: Assignment/Directory.Build.props (hunk 1 above)
Comment (ZackWard-student) [reply Assignment/Directory.Build.props:4]: Yup. They likely just missed the file it was in.

Comment (mmwoodfo) [reply Assignment/Directory.Build.props:4]: Yes - this should apply the defined properties to each project in the solution.

**Review by mmwoodfo — COMMENTED** [review]
Looks great.

**Pull request closed without merging**